SearchKit - Allow non-required select field to be blank

colemanw opened the pull request:
Overview
----------------------------------------
Fixes [dev/report#92](https://lab.civicrm.org/dev/report/-/issues/92)

Before
----------------------------------------
Cannot save empty value for non-required field using in-place edit.

After
----------------------------------------
Now you can.

colemanw committed:
SearchKit - Allow non-required select field to be blank

Fixes dev/report#92

demeritcowboy commented:
Just wondering if there's a way to do it more generically? Because it occurs to me it's not just dropdown fields - all the types need to take into account whether the field is required so we'd need to add something similar to all the .html files in that folder.

Having said that, this works for dropdowns and it should be a small edit to each file.

demeritcowboy commented:
Oh wait, this causes a problem for e.g. activity status. It shouldn't let you blank that out in a search display.

colemanw commented:
Ugh, I think it's not picking up the `required` attribute properly because the action is `get` not `create`. I'll work on that...

colemanw committed:
APIv4 - Don't suppress 'required' attribute in getFields

Before: APIv4 getFields would always set 'required' to FALSE unless the action was 'create'
After: Attribute is correctly passed through regardless of action

The original thinking behind this suppression was that only 'create' requires values,
and 'update' doesn't need anything but an ID.
However, that logic is enforced elsewhere, and doesn't need to be baked into getFields.
And sometimes it's useful to know if a field is required, e.g. in the SearchKit
in-place edit UI, it should be conveyed to the user if they are allowed to submit
an empty value for a given field.

colemanw commented:
This might do the trick @demeritcowboy

colemanw committed:
APIv4 - Set 'activity_type_id' to required

In APIv4 GetFields, sets activity field to required which should
not be null.

demeritcowboy commented:
not ok 24 - Error: api\v4\Action\ChainTest::testWithContactRef

Fatal error: Uncaught API_Exception: [0]: Mandatory values missing from Api4 Activity::create: status_id

  thrown in /home/jenkins/bknix-dfl/build/core-22330-4m123/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractCreateAction.php on line 63

colemanw commented:
Ok maybe I pushed it too far setting `status_id` to required. I see it shouldn't have to be, as the BAO will supply a default.
Seems like APIv4 getFields needs a way to distinguish `required` from `not null`, as the latter is more important for update actions.

demeritcowboy commented on CRM/Activity/DAO/Activity.php:
```diff
@@ -328,7 +328,8 @@ public static function &fields() {
           'import' => TRUE,
-          'where' => 'civicrm_activity.activity_type_id',
+          'where' => 'civicrm_activity.activity_type_id
+          ',
```
If a change is made to xml/schema/activity.xml, this change will get overwritten the next time GenCode is run to regenerate the DAO. I can't suggest something because I don't understand the purpose of this change - is something not adding whitespace when putting a query together and this works around that?

colemanw replied:
@demeritcowboy this change was actually produced by GenCode. I'm not sure why...

colemanw replied:
Ok, fixed it, the whitespace slipped in by mistake but simply running GenCode wouldn't fix it because the checksum doesn't consider whitespace.
Anyway, it's gone now.

demeritcowboy commented:
jenkins retest this please

colemanw commented:
jenkins retest this please

demeritcowboy commented on Civi/Api4/Service/Spec/Provider/ActivitySpecProvider.php:
```diff
@@ -23,39 +23,45 @@ class ActivitySpecProvider implements Generic\SpecProviderInterface {
+    // so activity type '1' doesn't necessarily exist. Best make the field required.
+    $spec->getFieldByName('activity_type_id')
+      ->setDefaultValue(NULL)
+      ->setRequired(TRUE);
```
I think `status` would need the same treatment? I can still blank it out in a search display.

colemanw replied:
Yea, that's what I meant by having a way for the api to understand `not null` separate from `required`. IMO we should merge this and then tackle that issue.

demeritcowboy commented:
And if I have a required custom field, it seems to let me blank it out.

colemanw commented:
Ooookay I think we need to take a step back and regroup. I'm going to close this.

demeritcowboy commented:
Ok. Thanks for working on it.

colemanw commented:
@demeritcowboy I decided to add a new property to getFields rather than fighting against the current `required` property which isn't the same thing as `not_null`. #22358

colemanw commented:
Also @demeritcowboy I thought one of the commits from this PR was still work keeping, if only on principle, so I split it to #22359
